### ISSUE TITLE
[BUG] Comparing decimal values with positive scale to MAX LONG or MIN LONG prodcues incorrect results

### ISSUE THREAD

**Issue by revans2** [rapidsai/cudf#9282]
**Describe the bug**
We have code to check if casting a Decimal to a long will overflow. The default code creates an INT64 scalar value with the min long value in it, and another with the max long value in it.  Then it uses binary ops like LESS, GREATER, LESS_EQUAL, and GREATER_EQUAL to see if the values are within the given range. This works generally great for bytes, shorts and ints, but for longs (INT64) it always comes back as true.  The values are in the desired range. I have to explicitly create a decimal scalar with a corresponding scale to make this check work. My guess is that both values are being cast to an INT64 before doing the comparison and the result by definition cannot be outside the range of those values.

**Comment by jrhemstad** [comment]
I'm not quite following what you mean. Can you give an example (just numbers, not code) of what you mean?

**Comment by revans2** [comment]
@jrhemstad

I have a DECIMAL64 value that is larger than can fit in an INT64.  A value of '999999999999999999' with scale 3 for example which is the equivalent of `999999999999999999000`.  Then I create a Scalar of INT64 with a value of `9223372036854775807` (`Long.MaxValue`). I ask cudf if my DECIMAL column > my INT64 scalar.  Logically it should say yes the decimal value is larger than the INT64 value, but cudf says false.  My guess is that cudf is casting the decimal to an INT64 first.  There is no int64 value that is > `Long.MaxValue`, so it will always return true.  This looks to be essentially the same kind of issue as was in the casting bug I filed, just in a separate place.

**Comment by jrhemstad** [comment]
Aha, that makes sense. Thanks.

**Comment by jrhemstad** [comment]
@codereport this sounds like a similar kind of problem?

**Comment by codereport** [comment]
> @codereport this sounds like a similar kind of problem?

This doesn't sound like the same thing. 

1. Comparing `INT64` and `Deimcal64` should not be possible
2. The correct way to do this is casting the `INT64` to `Decimal64` and then compare (ideally in the scale of the `Decimal64` number you are comparing it to)
3. If you only converted it to scale 0, I would expect the `Decimal64` number to potentially overflow, because the comparison operators are implemented using rescaling, similar to this:
```cpp
  auto const scale = std::min(lhs._scale, rhs._scale);
  return lhs.rescaled(scale)._value < rhs.rescaled(scale)._value;
```

**Note:** If you are currently able to compare `int64_t` with `decimal64` in libcudf, that might be a bug. While working on another job, I noticed that the compiled binary ops didn't have the same checks as the jit binary ops. I have fixed that in my other PR, but I haven't checked if that was introduced in my PR or in main.

**Comment by jrhemstad** [comment]
Hm, yeah, if the decimal64 number represents a value larger than INT64_MAX, how would you compare it to an int64? 

A quick and dirty way to do Bobby's check would be to convert the decimal64 to a `double` and then doing the comparison. You might get some weird effects due to the spacing between representable integers gets pretty far apart at that scale... I'm not sure if this would cause false negatives or false positives.

**Comment by jrhemstad** [comment]
> The correct way to do this is casting the INT64 to Decimal64 and then compare

Oh, right, this would work. So if you have `9223372036854775807`, you can create a `decimal64` with scale 3 to get the logical value `9223372036854775807000`. Then you can compare that against `999999999999999999000`. Since they have the same scale, it won't need to rescale either value and will just compare the rep between the two. 

Can't we do this automatically for binops? i.e., construct a decimal64 value around the integer value with the same scale as the other decimal value?

**Comment by revans2** [comment]
> The correct way to do this is casting the INT64 to Decimal64 and then compare

That is what I did as a work round in the Spark code.

**Comment by revans2** [comment]
> Am I missing a situation where this would be unsafe?
The first problem would show up for equality. `9223372036854775807` != `9223372036854775000` but if you automatically cast `9223372036854775807` to a DECIMAL64 with scale 3 you have lost some data and now it looks like they are equal. I think you can work around it and detect that you lost information when doing the cast to a decimal, and then always say that they are not equal.

The second issue would be with how you round.

For example: 

LONG `9223372036854775807` < DEC `9223372036854776000`

I would want to round the LONG towards 0 to get the right answer. But if I switch it (and adjust the DEC value so it shows the problem).

DEC `9223372036854775000` < LONG `9223372036854775807`

Now I want to round away from 0 for the long to give me the correct answer on the comparison.

I think that is correct.  Might need to play with some negative value examples to be sure that my rounding is spot on.

**Comment by codereport** [comment]
@jrhemstad 
> Oh, right, this would work. So if you have 9223372036854775807, you can create a decimal64 with scale 3 to get the logical value 9223372036854775807000. Then you can compare that against 999999999999999999000. Since they have the same scale, it won't need to rescale either value and will just compare the rep between the two.

This is correct, however, to be specific your value of `9223372036854775807000` is `9223372036854775807 * 1000` whereas the actual value will be `9223372036854775807 / 1000` and then `9223372036854775` is stored and it represents `9223372036854775000`.

> Can't we do this automatically for binops? i.e., construct a decimal64 value around the integer value with the same scale as the other decimal value? This would essentially be like implicit conversion for an integer when comparing against a float.

We could do this automatically. It is worth discussing though. I don't think we would want to auto-construct same-scale values when multiplying or adding - which means that the behaviour would be custom depending on the binary operation.

> In fact, can't we just add an overload for operator<(fixed_point, integer)? Am I missing a situation where this would be unsafe?

No, it isn't this simple. The `fixed_point` `operator<` is not what is invoked during a binary op. It would be a change somewhere else in the compiled binary_ops code.

**Comment by codereport** [comment]
@revans2 

> I think you can work around it and detect that you lost information when doing the cast to a decimal, and then always say that they are not equal.

Yea, I think it is possible for most case to construct to a scale n-1 and then do a `cudf::round`

> The second issue would be with how you round.
> DEC 9223372036854775000 < LONG 9223372036854775807

:arrow_up: this is the way it happens whenever `decimal` construction happens, truncation. This is the way CNL works and at one point we had rounding but it was decided to always truncate because of the complexity that it adds.

**Comment by revans2** [comment]
I get why you don't want the added complexity of round when constructing a decimal by default. But in this case we would probably want to construct the decimal with a scale of 0, and then explicitly round it to the appropriate scale using the correct rounding method based off of the operation being performed, along with all of the other gotchas involved with this type of an operation. There is a lot of very specific things that would need to be done correctly to make these comparison operations work correctly and generically. So I totally understand if the result of this BUG is simply CUDF does not support this type of operation and possibly some documentation why.
